### DISCUSSION TITLE
Hierarchy read API of collections in edit mode is failing with server error

### DISCUSSION THREAD

**Discussion by amitpriyadarshi** [Sunbird-Knowlg/Community#19] [Bugs]
@Krishnaj20 @vinukumar-vs @AmiableAnil 
Collection hierarchy read api is failing with server error with below response:

> {
  "id": "api.content.hierarchy.get",
  "ver": "3.0",
  "ts": "2022-06-08T11:21:25ZZ",
  "params": {
    "resmsgid": "b9352f77-a1c7-424b-baae-1b8b2d44f261",
    "msgid": null,
    "err": "ERR_SYSTEM_EXCEPTION",
    "status": "failed",
    "errmsg": "Something went wrong in the server while processing the request"
  },
  "responseCode": "SERVER_ERROR",
  "result": {}
}

**Analysis:**
There is an external property named `relational_metadata` introduced in `release-4.7.0` against the collection objects. If this property is `null` for a `Live` collection object and we call update hierarchy API, its draft version gets created in neo4j and the external property gets inserted into Cassandra against the draft object. In this case, the `relational_metadata` field gets set as `"null"` as a string (instead of `null`).

Because of which hierarchy read API throws an exception while parsing this value to object and throws Server Exception in response.

There are about 780 draft versions of collections that are impacted by this issue. This is impacting many users to update the already published collection.

**Comment by vinod303** [comment]
@amitpriyadarshi  i have also noticed the same issue

**Reply by anandvarada** [reply]
Hi @Krishnaj20 @vinukumar-vs @AmiableAnil 
We have users who are waiting for the issue fix as they are unable to edit / publish their collections, need a fix on priority

**Reply by kameshbhr** [reply]
@Krishnaj20 @vinukumar-vs This is a P1 issue that requires immediate hotfix as it is blocking DIKSHA users to update and publish any collection.

**Reply by Krishnaj20** [reply]
Team has started looking into it as a S1/P1 issue. Patch will be made available as soon as the root cause is identified and fixed. 

@vinukumar-vs

**Reply by Krishnaj20** [reply]
@anandvarada @amitpriyadarshi - From this description, the team is not able to understand the actual scenario which is causing this issue. Pls do add those details. 
@vinukumar-vs has set up a meeting to discuss more about this.

**Comment by Krishnaj20** [comment]
@kameshbhr & @anandvarada - This is happening only when Org Admin opens collections published by others from the 'All textbooks' tab. @amitpriyadarshi can correct the data for the impacted textbooks to unblock users. 
Please do note that this is not a recently introduced bug - it was there since 4.7 SunbirdEd release and the scenario to replicate this is also straight forward.  We've not seen this issue reported by the community (either by SunbirdEd (coKreat) or by the adopter QA team) when testing 4.7 or the subsequent releases. Therefore, it looks like a not so important workflow if it isn't part of the community's test scenarios. We will fix this issue in the current ongoing release (5.0).

**Reply by kameshbhr** [reply]
@anandvarada @amitpriyadarshi There is a disconnect between the analysis of DIKSHA team and what @Krishnaj20 has mentioned here. From what I see in DIKSHA, there are 3 P1 tickets raised which are related to the same issue. There is one ticket related to uploading content through VDN which also refers to the same issue.  So I am not sure why @Krishnaj20 is saying that this happens only in one specific scenario.
Let us do a triage if required. From my understanding, this is a P1 from DIKSHA side which has to be fixed as a hot fix.

**Reply by Krishnaj20** [reply]
Thanks @kameshbhr . 
@anandvarada - we were on a call with @amitpriyadarshi & Vinod and only for this scenario they were able to replicate this issue. Request you to add other scenarios as well.

**Reply by kameshbhr** [reply]
The three issues that I am referring to are:
1. Unable to edit Courses to add a Keyword
2. Unable to edit Textbooks
3. Unable to load textbooks in VDN project

**Reply by Krishnaj20** [reply]
@amitpriyadarshi - are you able to replicate the issue for the above mentioned scenarios?

**Reply by Jayaprakash8887** [reply]
@vinod303 Kindly check these scenarios as well and confirm.

**Reply by vinod303** [reply]
@Jayaprakash8887 @Krishnaj20 @kameshbhr  @anandvarada @vinukumar-vs 
For all the scenarios, the relational_metadata field gets set as "null" as a string, causing the issue.

For  
**2. Unable to edit Textbooks:**
 We are able to replicate as we did in the joint call, when the org admin open the content from the "All textbook" tab
 
 **3. Unable to load textbooks in VDN project**
 The problem as above. But the scenario here..
 The creator  of the textbook (who is also  a orgadmin), created a project adding these textbooks to the project and published it.
 Then after two days, he (ORGADMIN) opens the textbook via the "All textbook" tab. And  the reason mentioned by @Jayaprakash8887  and @amitpriyadarshi   the textbook did not open/open blank screen  due to server error.
Telemetry data also shows that, the user has  opened the  textbook from the "All Textbook" tab
So, post this,  when the user tried access  textbook via the project in VDN,   there is call from VDN to DIKSHA happens to fetch the textbooks, but due to the issue mentioned in the previous line, textbook will not load .

**1. Unable to edit Courses to add a Keyword**
 Here also, the issue is due to  relational_metadata field gets set as "null" as a string.  But, i am not able to find still in what scenario its happens for a course

**Reply by Jayaprakash8887** [reply]
@vinod303 @amitpriyadarshi Do you mean these flows were tested on a Draft textbook which has "null" relational_metadata? In that case, I think we have only 1 issue to address, right?

**Reply by Jayaprakash8887** [reply]
Patch for this issue is available as part of release-4.9.1 tracked under tickect https://project-sunbird.atlassian.net/browse/SB-30307

**Comment by Jayaprakash8887** [comment]
RCA: When the content-update (v3) API is invoked on a published collection, "relational_metadata" column in the cassandra is getting insert with null as string data for the newly created image node.

**Reply by vinukumar-vs** [reply]
@Krishnaj20 The fix is ready, we are verifying in dev. Once we verify we will push to preprod.
[JIRA ticket](https://project-sunbird.atlassian.net/browse/SB-30307)
